File: 1316 Supermarket/solution.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int p, d;
} Product;

int compare(const void *a, const void *b) {
    Product *p1 = (Product*)a;
    Product *p2 = (Product*)b;
    return p1->d > p2->d;
}

int main() {
    int n;
    while(scanf("%d", &n) == 1) {
        Product products[n];
        for(int i = 0; i < n; ++i)
            scanf("%d %d", &products[i].p, &products[i].d);
        qsort(products, n, sizeof(Product), compare);
        int *priorityQueue = NULL, size = 0;
        for(int i = 0; i < n; ++i) {
            if(products[i].d > size) {
                priorityQueue = (int*)realloc(priorityQueue, (size + 1) * sizeof(int));
                priorityQueue[size++] = products[i].p;
            } else {
                int minIndex = 0;
                for(int j = 1; j < size; ++j) {
                    if(priorityQueue[j] < priorityQueue[minIndex])
                        minIndex = j;
                }
                priorityQueue[minIndex] = priorityQueue[minIndex] > products[i].p ? priorityQueue[minIndex] : products[i].p;
            }
        }
        int profit = 0;
        for(int i = 0; i < size; ++i)
            profit += priorityQueue[i];
        printf("%d\n", profit);
        free(priorityQueue);
    }
	return 0;
}

File: 1316 Supermarket/solution.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StreamTokenizer st = new StreamTokenizer(br);
        StringBuilder output = new StringBuilder();
        while(st.nextToken() == StreamTokenizer.TT_NUMBER) {
    	    int n = (int)st.nval;
    	    Product[] products = new Product[n];
    	    for(int i = 0; i < n; ++i) {
    	        products[i] = new Product();
    	        st.nextToken();
    	        products[i].p = (int)st.nval;
    	        st.nextToken();
    	        products[i].d = (int)st.nval;
    	    }
    	    Arrays.sort(products, new Comparator<Product>() {
    	        @Override
    	        public int compare(Product p1, Product p2) {
    	            return Integer.compare(p1.d, p2.d);
    	        }
    	    });
    	    ArrayList<Integer> priorityQueue = new ArrayList<>();
    	    for(Product p : products) {
    	        if(p.d > priorityQueue.size())
    	            priorityQueue.add(p.p);
    	        else {
    	            int minIndex = 0;
    	            for(int j = 1; j < priorityQueue.size(); ++j) {
    	                if(priorityQueue.get(j) < priorityQueue.get(minIndex))
    	                    minIndex = j;
    	            }
    	            priorityQueue.set(minIndex, Math.max(priorityQueue.get(minIndex), p.p));
    	        }
    	    }
    	    int profit = 0;
    	    for(Integer i : priorityQueue)
    	        profit += i;
    	    output.append(profit).append('\n');
    	}
        System.out.print(output);
	}
}

class Product {
    int p, d;
}
